### PR TITLE
Support of an optional "payload" parameter

### PR DESCRIPTION
**Pull request by mamatsunami**
Appends an optional _payload_ parameter on **.can()** method and **condition callback passed to .allow()** method.

It makes it possible to check additional informations when it's time to decide if an action can be performed or not. For instance, on an update action, permissions can now be granted according to updated fields, as shown on the updated readme.

### CODE REVIEW

**Comment by dschnare** [comment]
This is a great addition. @vadimdemedes I think you should review this for sure.

File: test.js [hunk 1]
```diff
@@ -192,3 +192,24 @@ test('override instanceOf', t => {
 	t.false(cannot(user, 'read', product));
 	t.false(can(user, 'create', product));
 });
+
+test('allow can depend of an optional payload', t => {
+	const cancan = new CanCan();
+	const {can, allow} = cancan;
+
+	allow(User, 'update', User, (user, target, payload) => {
```
Comment (vadimdemedes) [inline test.js]: Could you clean this up? It's hard to follow.

File: test.js [hunk 2]
```diff
@@ -192,3 +192,24 @@ test('override instanceOf', t => {
 	t.false(cannot(user, 'read', product));
 	t.false(can(user, 'create', product));
 });
+
+test('allow can depend of an optional payload', t => {
```
Comment (vadimdemedes) [inline test.js]: => "pass options to the rule"

File: readme.md [hunk 3]
```diff
@@ -126,11 +132,24 @@ const post = new Post();
 can(user, 'view', post);
 ```
 
-### cannot(instance, action, target)
+```js
+/*
```
Comment (vadimdemedes) [inline readme.md]: Examples in readme should only contain code. If there are comments, they should be in text before that example.

File: readme.md [hunk 4]
```diff
@@ -126,11 +132,24 @@ const post = new Post();
...
+ * Authorization granted depending on updated fields
+ * with optional payload parameter
+ * /
+const admin = new User({roles: ['administrator']});
+const moderator = new User({roles: ['moderator']});
+
+can(admin, 'update', moderator, {fields: ["roles"]}); // true
```
Comment (vadimdemedes) [inline readme.md]: Use single quotes.

Comment (vadimdemedes) [reply readme.md]: And follow this format:

```js
can(...);
//=> return value
```

File: readme.md [hunk 5]
```diff
@@ -126,11 +132,24 @@ const post = new Post();
...
+can(moderator, 'update', moderator, {fields: ["roles", "username"]}); // false
+can(moderator, 'update', moderator, {fields: ["username"]}); // true
+```
+
+### cannot(instance, action, target[, payload])
 
 Inverse of `.can()`.
 
-### authorize(instance, action, target)
+### authorize(instance, action, target[, payload])
```
Comment (vadimdemedes) [inline readme.md]: `payload` => `options`.

File: index.js [hunk 6]
```diff
@@ -44,7 +44,7 @@ class CanCan {
 		});
 	}
 
-	can(performer, action, target) {
+	can(performer, action, target, payload) {
```
Comment (vadimdemedes) [inline index.js]: `payload` => `options`

**Comment by mamatsunami** [comment]
Hi @vadimdemedes !
I have been followed all your recommendations in (i hope) this last commit.
Thanks for your time and have a nice day !

**Comment by vadimdemedes** [comment]
Thanks for working on this!
